### PR TITLE
[libc++] Fix UTF-8 decoding in codecvts

### LINKED ISSUE #60177: [libc++] Codecvts that decode UTF-8 can incorrectly return partial instead of error
All codecvts that decode UTF-8 to other Unicode encodings (UTF-32, UTF-16 and UCS-2) have a bug in one case where they return `partial` instead of `error`. The case is when there is incomplete code sequence at the end of the input range but also in the visible trailing bytes there is some error. Here is example that reproduces the bug.

```cpp
#include <locale>
#include <cassert>

using namespace std;

int main()
{
	char in[]                 =  "bш\uAAAA\U0010AAAA";
	const char32_t expected[] = U"bш\uAAAA\U0010AAAA";
	char32_t out[ sizeof(expected)/sizeof(expected[0]) ];
	
	auto& loc = locale::classic();
	auto& cvt = use_facet<std::codecvt<char32_t, char, mbstate_t>>(loc);
	auto state = mbstate_t{};
	const char* in_next;
	char32_t* out_next;

	// \uAAAA is located at bytes in[3..5]
	auto orig = in[4];
	in[4] = 'z'; // we malform the second byte of the 3-byte sequence of \uAAAA

	// this call does not see the last byte of \uAAAA, it can only see its
	// first 2 bytes. If those 2 bytes were correct it should return partial,
	// but we malformed the second (and visible) byte. It should return error.
	auto res = cvt.in (state, in, in + 5, in_next, out, out + 5, out_next);
	assert(res == cvt.error); //error is correct, partial is returned

	in[4] = orig;

	// \U0010AAAA is located at bytes in[6..9]
	orig = in[7];
	in[7] = 'z'; // we malform the second byte of the 4-byte sequence of \U0010AAAA

	// this call does not see the last byte of \U0010AAAA, it can only see its
	// first 3 bytes. If those 3 bytes were correct it should return partial,
	// but we malformed the second (and visible) byte. It should return error.
	res = cvt.in (state, in, in + 9, in_next, out, out + 5, out_next);
	assert(res == cvt.error); //error is correct, partial is returned

	in[7] = orig;

	// can be also reproduced if we malform in[8]
}
```

### PR DESCRIPTION
**Pull request by dimztimz**
This patch fixes one case where the decoding member function `in()` was returning `partial` instead of `error`. Additionally, it adds large testsuite that tests all `codecvt` facets that were added in C++11 and in C++20. The testsuite covers this bug.

Fixes #60177.

### CODE REVIEW

**Comment by ldionne** [comment]
This used to be https://reviews.llvm.org/D143349.

@mordante can you take another quick look, I assume the feedback that had been given on Phab was applied here and this should be pretty much good to go.

**Comment by dimztimz** [comment]
> This used to be https://reviews.llvm.org/D143349.
> 
> @mordante can you take another quick look, I assume the feedback that had been given on Phab was applied here and this should be pretty much good to go.

It is not exactly the same it has even more tests that were not there previously, most notable it tests `codecvt_utf16`, too. That was not tested previously.

File: libcxx/test/std/localization/codecvt_unicode.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,2229 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// ADDITIONAL_COMPILE_FLAGS: -D_LIBCPP_DISABLE_DEPRECATION_WARNINGS
```
Comment (mordante) [inline libcxx/test/std/localization/codecvt_unicode.pass.cpp:9]: I just landed 233e7c5de2d2abda024577e79dcd195562de68c8
```suggestion
// ADDITIONAL_COMPILE_FLAGS: -D_LIBCPP_DISABLE_DEPRECATION_WARNINGS -D_LIBCPP_ENABLE_CXX26_REMOVED_CODECVT
```

**Review by mordante — APPROVED** [review]
Sorry the patch slipped under my radar. LGTM but one more change is required to make it work.
Please apply the change and rebase the patch to test the CI before landing.

**Comment by dimztimz** [comment]
CI is mostly successful. The few failures are not related to this patch.

**Comment by mordante** [comment]
> CI is mostly successful. The few failures are not related to this patch.

Thanks! For the future in GitHub you should be able to merge the changes yourself.

**Pull request merged**